File: runtime/Dart/lib/src/misc/pair.dart
/*
 * Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

import '../util/murmur_hash.dart';

class Pair<A, B> {
  final A a;
  final B b;

  const Pair(this.a, this.b);

  @override
  bool operator ==(other) {
    return identical(this, other) ||
        other is Pair && a == other.a && b == other.b;
  }

  @override
  String toString() {
    return '($a, $b)';
  }

  @override
  int get hashCode {
    MurmurHash.initialize();

    var hash = MurmurHash.initialize();
    hash = MurmurHash.update(hash, a);
    hash = MurmurHash.update(hash, b);
    return MurmurHash.finish(hash, 2);
  }
}

File: runtime/CSharp/Misc/Pair.cs
/* Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */
using System;

namespace Antlr4.Runtime.Misc
{
	public class Pair<A, B>
	{
		public readonly A a;
		public readonly B b;

		public Pair(A a, B b)
		{
			this.a = a;
			this.b = b;
		}

		public override bool Equals(Object obj)
		{
			if (obj == this)
			{
				return true;
			}
			else if (!(obj is Pair<A, B>))
			{
				return false;
			}

			Pair<A, B> other = (Pair<A, B>)obj;
			return (a == null ? other.a == null : a.Equals(other.a)) &&
				   (b == null ? other.b == null : b.Equals(other.b));
		}

		public override int GetHashCode()
		{
			int hash = MurmurHash.Initialize();
			hash = MurmurHash.Update(hash, a);
			hash = MurmurHash.Update(hash, b);
			return MurmurHash.Finish(hash, 2);
		}

		public override String ToString()
		{
			return String.Format("({0}, {1})", a, b);
		}
	}
}